dlt/destinations/impl/clickhouse/sql_client.py

```diff
@@ -0,0 +1,218 @@
+from contextlib import contextmanager
+from typing import (
+    Iterator,
+    AnyStr,
+    Any,
+    List,
+    Optional,
+    Sequence,
+    ClassVar,
+)
+
+import clickhouse_driver  # type: ignore[import-untyped]
+import clickhouse_driver.errors  # type: ignore[import-untyped]
+from clickhouse_driver.dbapi import OperationalError  # type: ignore[import-untyped]
+from clickhouse_driver.dbapi.extras import DictCursor  # type: ignore[import-untyped]
+
+import dlt
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.destinations.exceptions import (
+    DatabaseUndefinedRelation,
+    DatabaseTransientException,
+    DatabaseTerminalException,
+)
+from dlt.destinations.impl.clickhouse import capabilities
+from dlt.destinations.impl.clickhouse.configuration import ClickHouseCredentials
+from dlt.destinations.sql_client import (
+    DBApiCursorImpl,
+    SqlClientBase,
+    raise_database_error,
+    raise_open_connection_error,
+)
+from dlt.destinations.typing import DBTransaction, DBApi
+from dlt.destinations.utils import _convert_to_old_pyformat
+
+
+TRANSACTIONS_UNSUPPORTED_WARNING_MESSAGE = (
+    "ClickHouse does not support transactions! Each statement is auto-committed separately."
+)
+
+
+class ClickHouseDBApiCursorImpl(DBApiCursorImpl):
+    native_cursor: DictCursor
+
+
+class ClickHouseSqlClient(
+    SqlClientBase[clickhouse_driver.dbapi.connection.Connection], DBTransaction
+):
+    dbapi: ClassVar[DBApi] = clickhouse_driver.dbapi
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: ClickHouseCredentials) -> None:
+        super().__init__(credentials.database, dataset_name)
+        self._conn: clickhouse_driver.dbapi.connection = None
+        self.credentials = credentials
+        self.database_name = credentials.database
+
+    def has_dataset(self) -> bool:
+        return len(self._list_tables()) > 0
```

Comment:
way better would be to create "sentinel" table to indicate that dataset exists. please check waviate that does exactly that